### PR TITLE
user registration set up

### PR DESCRIPTION
**Pull request by danieljonesdmj**

### CODE REVIEW

File: lib/user.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    end
+  end
+
+  def self.add(username, password)
+    User.switch_database
+    result = @connection.exec("INSERT INTO users (username, password) VALUES ('#{username}', '#{password}') RETURNING id, username, password;")
+    User.new(result.first['id'], result.first['username'], result.first['password'])
+  end
+
+  def ==(username)
```
Comment (Ben-893) [inline lib/user.rb]: Why do we need this?? This is overriding inequality but it doesn't seem like the tests need it.

File: lib/user.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    @username == username.username
+  end
+
+  def self.switch_database
```
Comment (Ben-893) [inline lib/user.rb]: Should this be a private method? Same for the comment above (def ==(username))

Comment (Ben-893) [inline lib/user.rb]: Good to extract this to a new method.

File: lib/user.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    User.switch_database
+    result = @connection.exec("SELECT * FROM users")
+    result.map do |user|
+      User.new(user['id'], user['username'], user['password'])
+    end
+  end
+
+  def self.add(username, password)
+    User.switch_database
+    result = @connection.exec("INSERT INTO users (username, password) VALUES ('#{username}', '#{password}') RETURNING id, username, password;")
```
Comment (Ben-893) [inline lib/user.rb]: Need to encrypt the password before putting it into the database.

Comment (Ben-893) [reply lib/user.rb]: require 'bcrypt' gem in this file and gemfile

then you can do:

encrypted_password = BCrypt::Password.create(password)

then pass encrytped_password into the database

File: lib/user.rb (hunk 1 above)
Comment (laythq) [reply lib/user.rb]: Wrapping each PG object means the RSpec equality method doesnt work properly. Used the equality override to fix this. See userSpec.rb:7 expect(User.all).to include (user)

File: lib/user.rb (hunk 2 above)
Comment (laythq) [reply lib/user.rb]: Made private now, thanks

File: lib/user.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    User.new(result.first['id'], result.first['username'], result.first['password'])
+  end
+
+  def ==(username)
+    @username == username.username
+  end
+
+  def self.switch_database
+    if ENV['RACK_ENV'] == 'test'
```
Comment (Ben-893) [inline lib/user.rb]: Should this be replaced with ENV['ENVIRONMENT'] == 'test'?

File: lib/user.rb (hunk 2 above)
Comment (danieljonesdmj) [reply lib/user.rb]: switch database method is now private but equality method is public as it needs to be accessible for testing.

Comment (elishkaflint) [reply lib/user.rb]: ah yep - cool

File: spec/user_spec.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
+require 'user'
+
+describe User do
+  describe 'user.add' do
+    it 'creates a new user' do
+      users = User.add('username', 'password')
+      expect(User.all).to include(users)
+    end
+  end
+  describe 'user.authenticate' do
```
Comment (elishkaflint) [inline spec/user_spec.rb]: Please can we add a test that .authenticate doesn't work if the given password is incorrect?
In Chitter I had this:
```
it 'returns nil if the user gives the wrong email' do
  User.create('Chewbacca', 'chewbacca', 'chewbacca@gmail.com', '**pa$$w0rd4**')
  expect(User.authenticate('chewbacca@gmail.com', '**wrong password**')).to be_nil
end
```

Comment (danieljonesdmj) [reply spec/user_spec.rb]: added.

**Comment by elishkaflint** [comment]
:shipit:

**Pull request merged**